### DISCUSSION TITLE
Projects (not) showing effects, HUB75 vs WS2812

### DISCUSSION THREAD

**Discussion by b4wii** [PlummersSoftwareLLC/NightDriverStrip#357] [Q&A] [answered]
i do not have any effects after building for a basic esp32 dev board. a d1 mini. using ledstriplite as the default_envs in platformio.ini. 

i just have my led strip lighting up every 10th led statically. 
what am i doing wrong?

**Comment by davepl** [comment, marked as answer]
Is that the LEDSTRIP config?  If so, it does not compile with any built in effects.  It displays a status effect showing if its ready to receive WiFi Color Data:

Red: Waiting for WiFi
Green: Have WiFi, waiting for CLOCK
White: READY

To add another effect, you’d have to put it in the table in effects.cpp.  Look for the section labelled LEDSTRIP and steal a rainbow effect or similar from another section, add it there, and try it out!

- Dave

**Comment by Juris3D** [comment]
Hello, b4wii. May I ask, where did you take info on connections from esp32 to LED matrix? I thing I am running around in circles here, but I don't find schematics, or gerber file, or just textual descriptions what ports/pins of esp32 goes where. Could you please give some light on hardware side? Thank You in advance.

**Comment by davepl** [comment]
For LEDSTRIP, it’s nominally PIN5 on the ESP32.  You can change it in the code by changing the definition of LED_PIN0

**Comment by Juris3D** [comment]
Oh, but I am looking for connections for this one, using HUB75 display: https://youtu.be/X3V4gxd20FM
Or, that projects specs are not released yet at all? See, this is what confuses me. Thank You for help!

**Reply by b4wii** [reply]
Most esp32 development boards should have the pins labeled. I have purchased the esp32 d1 mini development board and it has the pinouts of what pin is what silkscreened on the back. The bare esp32 chip does not have them labeled. Also you would need some extra hardware to program the bare esp32 module if you don't buy a dev board with a USB connection

**Reply by b4wii** [reply]
Here is a link to what I'm using. https://www.aliexpress.us/item/2251832629731164.html

**Reply by b4wii** [reply]
Just looked up hub75. Unfortunately it looks like hub75 panels use a different protocol than what this project uses and so while you could hook up an esp32 to control a hub75 panel, this codebase will not work for you. This codebase is for ws2812b led based panels/light strips.

**Reply by Juris3D** [reply]
Thank You. But, I am talking about this project (link below), using HUB75 RGB LED Matrix. Is code and diagrams for it even available at all? 
Author's video: https://youtu.be/X3V4gxd20FM
HUB75 Matrix: https://www.aliexpress.com/item/32844117147.html

**Reply by rbergen** [reply]
You're talking about the Mesmerizer project. The source code for that is one of the projects in the NightDriverStrip (i.e. this repository's) codebase. 
The board in the video is also called Mesmerizer and was designed by @davepl himself. They can currently only be acquired from him directly.

**Reply by b4wii** [reply]
Ah yes I see. While he did make his matrix with a hub75 and created a custom interface board for it, he now has a Amazon link to recommended 16x16 ws2812b panels that you use three of to get the 16x32 size. https://amzn.to/3ABs5DK is his link to them. These work with the current code and allow for easier setup.

**Reply by Juris3D** [reply]
Thanks, guys... Wait, did "Mesmerizer" project move to WS2812 from HUB75? Or, "Mesmerizer" still stays on HUB75, but it is not openly available? Haha, see, there is still much to be confused about... :) 
P.S.: I am electronics engineer, I have experiences in this stuff, but, in this case, "I got lost" what project name and sources corresponds to what :))

**Reply by b4wii** [reply]
From the nightdriver GitHub project homepage he has links to different recommended hardware to get started with. He links to those 16x16 panels. If you wanted to play with the mesmerizer that uses the hub75 panel, you'll have to contact Dave to see about getting one of his boards that interfaces with it. Otherwise in the GitHub project it does have mesmerizer as an option for building the esp32 firmware which interfaces with his custom board.

**Reply by Juris3D** [reply]
Yes, I tried web uploader interface, and I did see mesmerizer as an option, so, it can be done, code-wise, what is still a mystery (for me) - pin connections from esp32 to HUB75 and pin(s) for audio inputs :) Again - Thank You for taking my calls on this! :)

**Reply by rbergen** [reply]
That's a good addition from @b4wii. Three of the WS2812 panels will actually get you to 16x48 pixels, which is what the Spectrum project expects. For me, that was quite easy to get to work with a M5StickC Plus. At about US$ 20 they're a bit more expensive than some of the bare dev boards, but they're very complete and easy to work with. 

About the names:

- The code in this repository (NightDriverStrip) includes a wide range of projects for a wide range of ESP32 chips/boards and LED devices. They are defined by a combination of configuration in platformio.ini and #defines in include/globals.h.
- The only project that currently uses the HUB75 is the Mesmerizer project, through the SmartMatrix library. Platformio.ini contains configurations to build the Mesmerizer project for the Mesmerizer board (as designed by @davepl) and the Wemos LOLIN D32 Pro - although I've never seen it run on the latter myself.
- The other "main" project that's set up to run on a matrix is Spectrum, which expects to be connected to 3 daisy-chained 16x16 WS2812 panels. The platformio.ini files contains configurations to build Spectrum on a number of different devices. The M5StickC Plus I mentioned earlier is one of them.

Pin configurations are generally defined in globals.h.

**Reply by b4wii** [reply]
Line 449 of globals.h for mesmerizer. It now uses code for ws2812b based matrixes and no longer supports hub75.

**Reply by rbergen** [reply]
Trust me, Mesmerizer really does use HUB75 matrices. If you look at the platformio.ini Mesmerizer environment configs, one of the defines that is set to 1 there is USE_MATRIX. They also pull in the SmartMatrix dependency. 

The USE_MATRIX define pulls in all kinds of stuff needed to interact with the HUB75 panel. For one, with that define set to 1 a completely different graphics class is compiled in. The image now uses LEDMatrixGFX (using SmartMatrix to control the HUB75) instead of the LEDStripGFX class that is used for WS2812B strips and matrices.

**Reply by Juris3D** [reply]
The plot thickens... :) I wonder what mister Plummer has to say on topic maybe? :)

**Comment by Juris3D** [comment]
rbergen - Thanks a Lot for your help, info, clarifications.

**Comment by robertlipe** [comment]
There's a lot of confusion in this thread.

This project supports HUB75 via the mesmerizer target. I'm currently
developing on that very combination, with an official Mesmerizer board.
While I've not been able to procure schematics, I don't think there's any
actual magic involved in mapping the five or sixe (or seven?) pins needed
to drive HUB75. (I've asked about schematics several times and have been
shut down every time, so I've resisted the urge to RE and publish the
pinouts....) In theory, via FastLED, it SHOULD support > 1 HUB75 panel,
though I've not explored that yet. The lower levels are meant to abstract
that.

Explore the MESMErIZER configuration in to explore the exact pins driven
and/or to change them to match your board. To ease your own life, I'd try
to wire your board to match MESMERIZER's hardware pinouts as the
configuration isn't really very flexible.

To the individual effects, the difference in WS2812 (probably APA, too) and
HUB75 is invisible. You set a pixel to a brightness, color, hue, etc. set
g()->leds[g()->xy(fooX, fooY) and the bulbs "magically" get set by
the lower layers. The FASTLED layer should abstract that away.

The details of actually configuring multiple matrices of 2812's and/or
HUB75's is probably messy. Even just changing to a 64x64 panel results in
breakage I've not yet tackled to fix. I suspect something isn't enabling
the E pin, but haven't traced it down.

As proof of concept, Here's (unsubmitted) effect # 193, "Eye Tunnel". The
flicker is camera frequency heterodyning and doesn't appear to a human
viewer. (Or so I'm told by a human.)

https://photos.app.goo.gl/32VBahYWMJQcUk1b6

Once I work out the schematics so I can drive a HUB75 and a few streams of
serial WS281x's from the same device, I'll be running a sanity check on the
effects in effects/matrix/*. They SHOULD work.

P.S. When referring to line numbers, please also refer to git tag levels.
Line numbers change.

**Comment by rbergen** [comment]
Dave will confirm that Mesmerizer uses a HUB75 panel, and the other projects address WS2812B strips and panels.

I know because I've been committing code to this repo for a while now. In fact, I opened a PR yesterday (#370) to make the separation between HUB75 ("matrix") code and WS2812B ("strip") code cleaner.

Here's a picture of my coffee table I just took. It's a bit overblown by the brightness of the WS2812B panels, but you can still see:
- 3 WS2812B panels connected to an M5StickC Plus, running the Spectrum project
- A HUB75 panel that has a Mesmerizer board attached to the back of it, running the Mesmerizer project.

![20230723_123634](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/10425218/256cf1e3-61bc-41f2-8f0a-4973fa7933e2)

**Comment by robertlipe** [comment]
I think we're agreeing, Rutger. I'm not questioning your obvious
credentials. I'm confirming that HUB75 does indeed work.

I previously questioned the clarity in this project between "Matrix"
meaning Hhub 5 vs. 2812 grids like you're showing. The code and the doc is
indeed full of confusion, IMO,  but I've not yet built up a single hardware
and code base that will run both at the same time. If the base code were to
do something like Marc's https://github.com/marcmerlin/SmartMatrix_GFX
where the base drivers are all transparent, the effects will Just Work on
boh of the displays you show. I haven't confirmed that yet. It's "just"
x/y/color/hue/brightness to be spun out to the hardware either way. In
theory, both can coexist. I don't know if they can both currently exist in
this code, but that IS in intent of this type of Adafruit-ish API.

We're agreeing: HUB75 absolutely works in this code in the mesmerizer
target.  There may be some later fluffing needed to get my new effects to
work on the 2812 grids, but I'm confident we can do that. Armed with some
diffusion filters to smooth out the giant "pixels" in the Neopixel
displays, they should look pretty good for the price. I have a basket ready
to wire up once I get a combo HUB75 + STRIP board going in my lab.

Blink on!

**Reply by rbergen** [reply]
No worries, @robertlipe.

For one, I wasn't responding to your message, but to @Juris3D's question what Dave would say. Our messages just crossed, if that's the right way to put it.

Anyway, we're coming from different angles but arrive at the same overall conclusion. So, all good as far as I'm concerned. :)

**Comment by robertlipe** [comment]
I have many, many feet of WS281x strips awaiting that day. Strips, grids,
rings, hexagons. They all "just" await time and hookup clarity. I have an
embarrassing many hundreds of feet and thousands of bulbs ranging from
720LED/M to 120LED/M and down to the 30LED/M range.

Once I move to strip mode, I can exercise a ridiculous amount of
configurations. It's just a matter of focus for now.

**Comment by Juris3D** [comment]
on a technical side of things, I am curious on one thing. Knowing that HUB75 does not do any PWM, does "mesmerizer" brute force writes a lot of "frames" (sorry my English, I am Latvian)? I mean, I was surprised on mesmerizer HUB75 videos, about smoothness of those.

**Reply by rbergen** [reply]
That's a very good question, actually. Dave in particular spends a lot of effort on trying to push up the panel refresh rate, optimize double buffering, etc. All without overloading the CPU - which we're regularly pretty close to.

So yes, in short. :)

**Comment by robertlipe** [comment]
Cool. That happens at 6 am. :-/

WIthout being rude to our host, I'm trying to break the mindset of "what
would Dave say". When a couple of people have clearly run a configuration,
it works. For this to be a successful open source project, it needs to rely
upon the trust built by others in the project and not rely upon
an individual. (In another project, I'm that individual, and would welcome
sharing that very hat...)  I'm underscoring your answer that This Works,
contrary to earlier assertions. We're cool here.

This is now way OT for the original bugreport here. If there's an actual
issue left, it should probably go to another report and let this one run
its course.

GitHub bugreports, like many others, suffer both when contributors are too
frequent (you and I crossing paths) and too infrequent (unanswered
reports). There just isn't a "right" solution to that, IME.

RJL

**Reply by Juris3D** [reply]
While I understand issue of going "off-topic", see it this way: this thread now is THE best source on any clarification on all mesmerizer/nightdriver/hub75/ws2812 - thing.

**Reply by rbergen** [reply]
@robertlipe I think your comment lightly touches upon some meta-topics about this project that we could discuss at length, but I think that should (indeed) better be done in a separate discussion.

I will respond to the point of order about going off-topic, because I consider that feedback on the way I'm managing this repository.
Had this been an actual bug report (i.e. an Issue), then I would have managed it more tightly to retain (or increase) focus. This being a Q&A discussion, I allow more room for conversations to waver, Sometimes good things come from it, as I think happened in this case.
With this discussion now containing what it does, I think it is appropriate to change its title. I will do so later today.

**Comment by robertlipe** [comment]
Agreed. I'll take carefully worded Latvian (or anything else that Google
Translate can handle) over sloppy English every day. (I speak only English.)

HUB75 boards and WS281x both take a lot of hand-holding at the frame
level.  Effects like to plop pixels around, but HUB75 needs refreshed crazy
frequently as the whole panel refreshes and WS281x needs nutty hand-holding
at the sub-pixel level to fit into SPI or other PWM context to fit into a
grid.

FWIW, I've now converted/created over 60 HUB75 effects[1] with Mesmerizer
hardware and a 64x32 display and only really wished for a higher frame rate
2-3 times. For the most part, it works with sufficient headroom. If we were
running BOTH HUB75 and WS2812, life might be tough. Feeding WS2812 involves
all kinds of trickery on ping-pong DMA of SPI frames and such. They're hard
to keep fed without slapping the host CPU silly, but done right, can get
down into the sub 5% range for fixed cases.

[1] Not yet submitted for review/integration. Meant only to show real-world
experience.

**Comment by robertlipe** [comment]
Agreed all around.

Well done.
